File: 2021/day05/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type Line struct {
	x1, y1, x2, y2 int
}

type Point struct {
	x, y int
}

func parseInput() ([]Line, error) {
	exePath, err := os.Executable()
	if err != nil {
		exePath, _ = os.Getwd()
	} else {
		exePath = filepath.Dir(exePath)
	}

	// Try multiple paths
	paths := []string{
		filepath.Join(exePath, "..", "input.txt"),
		filepath.Join(".", "..", "input.txt"),
		"../input.txt",
	}

	var file *os.File
	for _, p := range paths {
		file, err = os.Open(p)
		if err == nil {
			break
		}
	}
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []Line
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		parts := strings.Split(line, " -> ")
		start := strings.Split(parts[0], ",")
		end := strings.Split(parts[1], ",")

		x1, _ := strconv.Atoi(start[0])
		y1, _ := strconv.Atoi(start[1])
		x2, _ := strconv.Atoi(end[0])
		y2, _ := strconv.Atoi(end[1])

		lines = append(lines, Line{x1, y1, x2, y2})
	}

	return lines, scanner.Err()
}

func sign(x int) int {
	if x > 0 {
		return 1
	} else if x < 0 {
		return -1
	}
	return 0
}

func countOverlaps(lines []Line, includeDiagonals bool) int {
	grid := make(map[Point]int)

	for _, l := range lines {
		dx := sign(l.x2 - l.x1)
		dy := sign(l.y2 - l.y1)

		// Skip diagonals in part 1
		if !includeDiagonals && dx != 0 && dy != 0 {
			continue
		}

		x, y := l.x1, l.y1
		for {
			grid[Point{x, y}]++
			if x == l.x2 && y == l.y2 {
				break
			}
			x += dx
			y += dy
		}
	}

	count := 0
	for _, v := range grid {
		if v >= 2 {
			count++
		}
	}
	return count
}

func part1(lines []Line) int {
	return countOverlaps(lines, false)
}

func part2(lines []Line) int {
	return countOverlaps(lines, true)
}

func main() {
	lines, err := parseInput()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Part 1: %d\n", part1(lines))
	fmt.Printf("Part 2: %d\n", part2(lines))
}

File: 2021/day05/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

function parseInput() {
  return input.split('\n').map(line => {
    const [start, end] = line.split(' -> ');
    const [x1, y1] = start.split(',').map(Number);
    const [x2, y2] = end.split(',').map(Number);
    return { x1, y1, x2, y2 };
  });
}

function sign(x) {
  if (x > 0) return 1;
  if (x < 0) return -1;
  return 0;
}

function countOverlaps(lines, includeDiagonals = false) {
  const grid = new Map();

  for (const { x1, y1, x2, y2 } of lines) {
    const dx = sign(x2 - x1);
    const dy = sign(y2 - y1);

    // Skip diagonals in part 1
    if (!includeDiagonals && dx !== 0 && dy !== 0) {
      continue;
    }

    let x = x1, y = y1;
    while (true) {
      const key = `${x},${y}`;
      grid.set(key, (grid.get(key) || 0) + 1);
      if (x === x2 && y === y2) break;
      x += dx;
      y += dy;
    }
  }

  let count = 0;
  for (const v of grid.values()) {
    if (v >= 2) count++;
  }
  return count;
}

function part1() {
  const lines = parseInput();
  return countOverlaps(lines, false);
}

function part2() {
  const lines = parseInput();
  return countOverlaps(lines, true);
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());
